src/modules/automata.rs

```diff
@@ -99,20 +100,35 @@ impl Automata {
     }
 
     pub fn full_match(&self, expr: &str) -> bool {
-        if let Some(matched) = self.greedy_search_impl(expr, true) {
-            matched.len() == expr.len()
-        } else {
-            false
+        match self.greedy_search(expr) {
+            Ok(result) => {
+                if let Some(matched) = result {
+                    matched.len() == expr.len()
+                } else {
+                    false
+                }
+            }
+            Err(error) => {
+                println!("Warning: {error:#?}");
+                false
+            }
         }
     }
 
-    pub fn greedy_search(&self, expr: &str) -> Option<String> {
-        self.greedy_search_impl(expr, false)
+    pub fn greedy_search(&self, expr: &str) -> Result<Option<String>, Error> {
+        let mut search_results = self.search(expr, true);
+
+        if search_results.len() > 1 {
+            Err(Error::from("too many results returned for greedy search"))
```

Comment:
Since search is a public interface, I think it would also be an error if search(expr, true) returned more than one value. So I think it would make more sense to have the check (if you must have one).

Personally, I feel that this is something that should be handled by tests. If a bug prevents you from proceeding, I can understand emitting an error. But here we have a return value, and I don't think its worth sanity checking it before returning.

I'm also not a fan of exposing two ways of doing the same thing. Sometimes it makes sense. I assume you kept greedy_search as a convenient way of getting the single value, instead of the vector of size 1. In that case, maybe it would make sense to make `search` private, and expose a `global_search` function instead?

Reply:
This is currently outdated, but: initially I meant to make `search` the "global search" function (or rather, with passing `greedy = false`  to `search` means running a "global search"), which specialises to `greedy_search` (using `greedy = true`). I think what you suggested is probably cleaner though, so now `search` is private and we expose `global_search` instead.